File: automaton-maui/Core/Simulator.cs
using AutomatonDesigner.Models;

namespace AutomatonDesigner.Core;

/// <summary>
/// Pure simulation engine for DFA, NFA, and ε-NFA automata.
/// </summary>
public static class Simulator
{
    /// <summary>Compute the epsilon closure of a set of states.</summary>
    public static HashSet<State> EpsilonClosure(Automaton aut, HashSet<State> initial)
    {
        var visited = new HashSet<State>(initial);
        var frontier = new HashSet<State>(initial);

        while (frontier.Count > 0)
        {
            var next = new HashSet<State>();
            foreach (var t in aut.Transitions)
            {
                if (t.Label is OnEpsilon && frontier.Contains(t.From) && !visited.Contains(t.To))
                    next.Add(t.To);
            }
            if (next.Count == 0) break;
            visited.UnionWith(next);
            frontier = next;
        }
        return visited;
    }

    /// <summary>Advance one step: compute next states after reading a symbol.</summary>
    public static HashSet<State> Step(Automaton aut, HashSet<State> current, Symbol sym)
    {
        var targets = new HashSet<State>();
        foreach (var t in aut.Transitions)
        {
            if (t.Label is OnSymbol s && s.Symbol == sym && current.Contains(t.From))
                targets.Add(t.To);
        }
        return aut.Type == AutomatonType.ENFA ? EpsilonClosure(aut, targets) : targets;
    }

    /// <summary>Compute initial active states (with ε-closure for ε-NFA).</summary>
    static HashSet<State> InitialStates(Automaton aut)
    {
        var init = new HashSet<State> { aut.Start };
        return aut.Type == AutomatonType.ENFA ? EpsilonClosure(aut, init) : init;
    }

    /// <summary>Run a full simulation and return step-by-step trace.</summary>
    public static SimResult Simulate(Automaton aut, string input)
    {
        var active = InitialStates(aut);
        var steps = new List<SimStep>
        {
            new(0, null, new HashSet<State>(active), input)
        };

        for (int i = 0; i < input.Length; i++)
        {
            var sym = new Symbol(input[i]);
            active = Step(aut, active, sym);
            steps.Add(new SimStep(i + 1, sym, new HashSet<State>(active), input[(i + 1)..]));
        }

        bool accepted = active.Overlaps(aut.Accept);
        return new SimResult(accepted, steps, input);
    }

    /// <summary>Quick check: does the automaton accept the string?</summary>
    public static bool Accepts(Automaton aut, string input) =>
        Simulate(aut, input).Accepted;

    /// <summary>Run a batch of test cases.</summary>
    public static List<TestResult> RunTests(Automaton aut, List<TestCase> tests) =>
        tests.Select(tc =>
        {
            bool actual = Accepts(aut, tc.Input);
            return new TestResult(tc, actual == tc.Expected, actual);
        }).ToList();
}

File: automaton-designer/src/Automaton/Simulator.hs
{-# LANGUAGE OverloadedStrings #-}

-- | Pure simulation engine for automata.
--
-- Supports DFA (single active state), NFA (set of active states),
-- and ε-NFA (epsilon closure computed at each step).
module Automaton.Simulator
  ( simulate
  , simulateStep
  , epsilonClosure
  , step
  , accepts
  , runTests
  ) where

import Data.Set  (Set)
import qualified Data.Set  as Set
import Data.Text (Text)
import qualified Data.Text as T

import Automaton.Types

-- ---------------------------------------------------------------------------
-- Epsilon closure
-- ---------------------------------------------------------------------------

-- | Compute the epsilon closure of a set of states.
--
-- Returns all states reachable from the given set via zero or more
-- ε-transitions (fixed-point iteration).
epsilonClosure :: Automaton -> Set State -> Set State
epsilonClosure aut initial = go initial initial
  where
    go visited frontier
      | Set.null newReachable = visited
      | otherwise = go visited' newReachable
      where
        -- States reachable via one ε-step from the frontier
        reachable = Set.fromList
          [ transTo t
          | t <- Set.toList (automTransitions aut)
          , transFrom t `Set.member` frontier
          , transLabel t == OnEpsilon
          ]
        newReachable = Set.difference reachable visited
        visited'     = Set.union visited newReachable

-- ---------------------------------------------------------------------------
-- Single-step transition
-- ---------------------------------------------------------------------------

-- | Advance one step: given the current set of active states and a symbol,
-- compute the next set of active states.
--
-- For ε-NFA, epsilon closure is applied after following symbol-transitions.
step :: Automaton -> Set State -> Symbol -> Set State
step aut current sym =
  let directTargets = Set.fromList
        [ transTo t
        | t <- Set.toList (automTransitions aut)
        , transFrom t `Set.member` current
        , transLabel t == OnSymbol sym
        ]
  in case automType aut of
       ENFA -> epsilonClosure aut directTargets
       _    -> directTargets

-- ---------------------------------------------------------------------------
-- Full simulation
-- ---------------------------------------------------------------------------

-- | Compute initial active states (applying ε-closure for ε-NFA).
initialStates :: Automaton -> Set State
initialStates aut = case automType aut of
  ENFA -> epsilonClosure aut (Set.singleton (automStart aut))
  _    -> Set.singleton (automStart aut)

-- | Run a complete simulation and return a step-by-step trace.
simulate :: Automaton -> Text -> SimResult
simulate aut input =
  let symbols = map Symbol (T.unpack input)
      initActive = initialStates aut

      -- Step 0: before reading any input
      step0 = SimStep
        { stepNumber       = 0
        , stepSymbol       = Nothing
        , stepActiveStates = initActive
        , stepRemaining    = input
        }

      -- Build all subsequent steps with scanl
      allSteps = scanl advance step0 (zip [1..] symbols)

      finalActive = stepActiveStates (safeLast step0 allSteps)
      accepted    = not $ Set.null $ Set.intersection finalActive (automAccept aut)
  in SimResult
    { simAccepted = accepted
    , simSteps    = allSteps
    , simInput    = input
    }
  where
    advance prev (i, sym) =
      let nextStates = step aut (stepActiveStates prev) sym
      in SimStep
        { stepNumber       = i
        , stepSymbol       = Just sym
        , stepActiveStates = nextStates
        , stepRemaining    = T.drop i input
        }

    safeLast def [] = def
    safeLast _   xs = last xs

-- | Perform a single simulation step from a given set of active states.
-- Useful for step-by-step interactive simulation.
simulateStep :: Automaton -> Set State -> Symbol -> (Set State, Bool)
simulateStep aut current sym =
  let next     = step aut current sym
      accepted = not $ Set.null $ Set.intersection next (automAccept aut)
  in (next, accepted)

-- | Quick check: does the automaton accept the given string?
accepts :: Automaton -> Text -> Bool
accepts aut = simAccepted . simulate aut

-- ---------------------------------------------------------------------------
-- Batch test runner
-- ---------------------------------------------------------------------------

-- | Run a list of test cases against an automaton.
runTests :: Automaton -> [TestCase] -> [TestResult]
runTests aut = map runOne
  where
    runOne tc =
      let actual = accepts aut (testInput tc)
      in TestResult
        { trTestCase = tc
        , trPassed   = actual == testExpected tc
        , trActual   = actual
        }
